drivers: sensor: add support for sensirion shtc3

str4t0m opened the pull request:
Add support for sensirion shtcx temperature and humidity sensors.
Tested with shtc3, but shtc1 has the same interface with the exception that it is missing the sleep and wakeup commands.


Additionally the sht3x driver is refactored:
- replaces custom crc8 with sys/crc8
- uses sys_put/sys_get helpers for byteorder specific operations instead of shifted bytes.

str4t0m committed:
tests: drivers: enable shtc3 build

Enable shtc3 sensor for build_all test

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

mbolivar-nordic commented on drivers/sensor/shtcx/Kconfig:
```diff
@@ -0,0 +1,26 @@
+	prompt "Measurement mode"
+	default SHTCX_MODE_NORMAL
+
+config SHTCX_MODE_NORMAL
+	bool "normal mode"
+
+config SHTCX_MODE_LOW_POWER
+	bool "low power mode"
```
I'm just guessing, but it looks this should be in devicetree instead.

str4t0m replied:
I have moved the selection of the measurement mode to devicetree.

mbolivar-nordic commented on drivers/sensor/shtcx/shtcx.c:
```diff
@@ -0,0 +1,276 @@
+		.bus = DEVICE_DT_GET(DT_INST_BUS(inst)),		\
+		.base_address = DT_INST_REG_ADDR(inst),			\
+		.chip = TO_CHIP(inst),					\
+		.clock_stretching =					\
+			DT_INST_PROP_OR(inst, clock_stretching, 0U),	\
```
Just use DT_INST_PROP. It does the right thing for booleans. See the documentation for details.

str4t0m replied:
done.

mbolivar-nordic reviewed (changes requested):
Looks good. One DT API nit and a thought about whether a Kconfig option should be in DT.

str4t0m commented:
rebased to resolve conflict.

str4t0m commented:
fixed a rounding error at raw conversion, by switching from shifts to divisions.

mbolivar-nordic commented on drivers/sensor/shtcx/shtcx.c:
```diff
@@ -0,0 +1,279 @@
+			return -EINVAL;
+		}
+		shtcx_sleep(dev);
+	} else {
+
+	}
```
Is something missing here?

If not, the empty block should be removed.

str4t0m replied:
Usually I don't use this style, but saw #33198 fixing guide line 15.7, so I thought i could use this style.
I have updated it to use two separate if blocks.

mbolivar-nordic commented on drivers/sensor/shtcx/shtcx.c:
```diff
@@ -0,0 +1,279 @@
+		return -ENODEV;
+	}
+
+	if (!cfg->base_address) {
+		LOG_ERR("No I2C address");
+		return -EINVAL;
+	}
```
This does not seem needed. If someone puts 0 in the devicetree `reg` for this node, that is a simple misconfiguration, no? What am I missing?

str4t0m replied:
removed

mbolivar-nordic commented on drivers/sensor/shtcx/shtcx.h:
```diff
@@ -0,0 +1,86 @@
+struct shtcx_sample {
+	uint16_t temp;
+	uint16_t humidity;
+} __packet;
```
Does the `build_all` change really build this? It looks like this should cause a build error because it should be `__packed`, not `__packet`.

str4t0m replied:
Thanks! Good find. Using west and the zephyr-sdk I didn't get a warning, and as the members were correctly aligned on my test platform I didn't notice it.
After correcting it to __packed, the struct was aligned to an odd address and curiously the mcu got stuck at the first write to it. Therefore I have additionally added __aligned(2). So this should work on all platforms now.

mbolivar-nordic commented:
Thanks for the update! I noticed a few things.

str4t0m commented:
Thanks for the review, I have addressed your findings.

mbolivar-nordic reviewed (approved):
Device/DT bits lgtm

jfischer-no commented on drivers/sensor/shtcx/shtcx.c:
```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright (c) 2021 Thomas Stranger
```
`* Copyright (c) 2021 Thomas Stranger <thomas.stranger@outlook.com>` ?

str4t0m replied:
Sorry to bug you. I have seen many license headers w/o the mail.
Is it mandatory for private contributors to include the mail in each license header?

jfischer-no commented on drivers/sensor/shtcx/shtcx.c:
```diff
@@ -0,0 +1,273 @@
+};
+
+/*
+ * CRC algorithm parameters were taken from the
```
Please use [crc8()](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/sys/crc.h#L59)

str4t0m replied:
done

jfischer-no commented on drivers/sensor/shtcx/shtcx.c:
```diff
@@ -0,0 +1,273 @@
+{
+	if (shtcx_write_command(dev, SHTCX_CMD_SLEEP) < 0) {
+		return -EIO;
+	}
```
Add empty line between } and return line in the functions above.
Same for L:145, L:248

str4t0m replied:
done

str4t0m commented:
I have added the empty lines and replaced the crc8 implementation.
Added an additional commit that replaces the custom crc8 implementation for the sensirion sht3xd driver.
Please see comment with a question regarding the license header.

str4t0m commented:
As suggested I have rebased and now all the clang-build failures are gone.

str4t0m commented:
I believe there shouldn't be any issues left, can this be merged within the v2.6.0 merge window?

jfischer-no commented on samples/sensor/shtcx/src/main.c:
```diff
@@ -0,0 +1,54 @@
+		return;
+	}
+
+	struct sensor_value temp, humidity;
```
please move it to the top of the block.

str4t0m replied:
Done.

jfischer-no commented on drivers/sensor/shtcx/shtcx.h:
```diff
@@ -0,0 +1,86 @@
+/* Soft reset time is 230us for shtc1 and 240us for shtc3 */
+#define SHTCX_SOFT_RESET_TIME_US 240U
+
+#define SHTCX_MAX_READ_LEN 6
```
Please indent/align it whit tabs right away.

str4t0m replied:
Done.

str4t0m commented on drivers/sensor/sht3xd/sht3xd.c:
```diff
@@ -43,22 +44,8 @@ static const int measure_wait[3] = {
 static uint8_t sht3xd_compute_crc(uint16_t value)
 {
 	uint8_t buf[2] = { value >> 8, value & 0xFF };
```
Done.
As you suggested to squash the "crc8 commit" for the sht3xd driver, I have done this refactoring also in the same commit.
But as this is not dependent on the new driver it would probably be better to do this refactoring in a separate commit,
wouldn't it?

str4t0m commented on drivers/sensor/sht3xd/sht3xd.c:
```diff
@@ -11,6 +11,7 @@
 #include <kernel.h>
 #include <drivers/sensor.h>
 #include <sys/__assert.h>
+#include <sys/crc.h>
```
Done.

jfischer-no replied:
Sorry, I probably missed something. Changes in the sht3xd driver should be in separate commit/PR. Your PR is about shtc1 and shtc3 driver, to see sht3xd changes here is a bit confusing.

str4t0m commented:
Thanks for the review I have addresses these changes.
Additionally I have replaced the device_pm_control_nop with NULL, to get rid of the deprecation warning that emerged in the meantime.

str4t0m committed:
drivers: sensor: shtcx: add driver supporting shtc1 and shtc3

Add driver for sensirion consumer humidity sensor line.
Supports shtc1 and shtc3, but only shtc3 is tested.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
drivers: sensor: sht3xd: use sys functions for crc and byteorder

replace custom crc8 with sys/crc8
use sys_put/sys_get helpers for byteorder specific operations

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

jfischer-no commented on drivers/sensor/shtcx/shtcx.c:
```diff
@@ -0,0 +1,263 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT sensirion_shtcx
```
Did you lose shtcx bindings during rebasing?

str4t0m replied:
Indeed, thanks. This is also the cause of the failing ci tests.

str4t0m commented:
Since the last reviews I have dropped the commit with the sample.
Furthermore the way I assigned the devicetree `chip` and `measure-mode` enums did no longer work so I have made minor changes. (Initially I used the enum token now I am using the enum idx).

Could you please have another look?
